// testcafe-browser-provider-browserstack/lib/utils/request-api.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const pinkie_1 = __importDefault(require("pinkie"));
const request_promise_1 = __importDefault(require("request-promise"));
const ERROR_MESSAGES = __importStar(require("../templates/error-messages"));
const apiRequestPromise = pinkie_1.default.resolve(null);
function default_1(apiPath, params = {}) {
    if (!process.env['BROWSERSTACK_USERNAME'] || !process.env['BROWSERSTACK_ACCESS_KEY'])
        throw new Error(ERROR_MESSAGES.BROWSERSTACK_AUTHENTICATION_FAILED());
    var { body, executeImmediately } = params, queryParams = __rest(params, ["body", "executeImmediately"]);
    var opts = {
        url: apiPath.url,
        auth: {
            user: process.env['BROWSERSTACK_USERNAME'],
            pass: process.env['BROWSERSTACK_ACCESS_KEY'],
        },
        headers: {
            'user-agent': 'testcafe-browserstack',
        },
        qs: Object.assign({}, queryParams),
        method: apiPath.method || 'GET',
        json: apiPath.encoding === void 0
    };
    const proxy = process.env['BROWSERSTACK_PROXY'];
    if (proxy)
        opts.proxy = `http://${proxy}`;
    if (body)
        opts.body = body;
    if (apiPath.encoding !== void 0)
        opts.encoding = apiPath.encoding;
    const chainPromise = executeImmediately ? pinkie_1.default.resolve(null) : apiRequestPromise;
    const currentRequestPromise = chainPromise
        .then(() => request_promise_1.default(opts))
        .catch(error => {
        if (error.statusCode === 401)
            throw new Error(ERROR_MESSAGES.BROWSERSTACK_AUTHENTICATION_FAILED());
        throw error;
    });
    return currentRequestPromise;
}
exports.default = default_1;
